### DISCUSSION TITLE
Used custom Tokenizer but not recognizing entity

### DISCUSSION THREAD

**Discussion by farrandi** [explosion/spaCy#8390] [Help: Coding & Implementations] [answered]
The current tokenizer in spacy (English) does not separate this:
```Python:
nlp = spacy.load('en_core_web_sm')
text = "LV FS:36.3%"
doc = nlp(text)
print([token for token in doc])
```
This gives an output of `[LV, FS:36.3, %]`

So I created a custom tokenizer where I modified the infix and I trained a spaCy model with this custom tokenizer following the documentation on the spaCy website.

The tokenizer works as shown in the code below:
```Python:
nlp = spacy.load(r"./output/model-best")
doc = nlp("LV FS:36.3%")
print([t for t in doc])
spacy.displacy.render(doc, style = "ent", jupyter = True)
```
This gives the output: 
![image](https://user-images.githubusercontent.com/54917098/121943696-b1b72600-cd06-11eb-97a6-03e3d04db3d2.png)

However, I would like it to be 'LV FS' (Observation) : '36.3'(Value) '%'(Units)

Is there a reason it still does not recognize it as such?

**Comment by polm** [comment, marked as answer]
Without seeing your training data I don't think we can say whether it should learn that or not. If your annotations are consistent with your new tokenizer then it should certainly be able to learn the labels.

Can you clarify what version of spaCy you're using?

**Reply by farrandi** [reply]
This is my current spacy version:
```
============================== Info about spaCy ==============================

spaCy version    3.0.6
Location         /home/farrandi/.local/lib/python3.8/site-packages/spacy
Platform         Linux-5.4.72-microsoft-standard-WSL2-x86_64-with-glibc2.29
Python version   3.8.5
Pipelines        en_core_web_sm (3.0.0), en_core_web_lg (3.0.0), en_core_sci_sm (0.4.0)
```

My annotations seem to be consistent from looking at the labels. It is important to note that I used another program (called Doccano) for the labels. I also used training data of 2 pages of text instead of sentences or paragraphs. I did read discussion [#8370](https://github.com/explosion/spaCy/discussions/8370)  so is it maybe because of this? since splitting the long texts does not affect the ability for us to annotate. Thanks!

**Reply by polm** [reply]
Using Doccano shouldn't be an issue. 

Using 2 pages of text sounds kind of strange. How many annotations do you have? How many documents is it? What is the output of `spacy debug data` on your data?

**Reply by farrandi** [reply]
Here are the results of my spacy debug data:
```
============================ Data file validation ============================
✔ Corpus is loadable
✔ Pipeline can be initialized with data

=============================== Training stats ===============================
Language: en
Training pipeline: tok2vec, ner
80 training docs
20 evaluation docs
✔ No overlap between training and evaluation data
✘ Low number of examples to train a new pipeline (80)

============================== Vocab & Vectors ==============================
ℹ 78478 total word(s) in the data (4654 unique)
ℹ No word vectors present in the package

========================== Named Entity Recognition ==========================
ℹ 12 label(s)
0 missing value(s) (tokens with '-' label)
⚠ 103 entity span(s) with punctuation
✔ Good amount of examples for all labels
✔ Examples without occurrences available for all labels
✔ No entities consisting of or starting/ending with whitespace
Entity spans consisting of or starting/ending with punctuation can not be
trained with a noise level > 0.

================================== Summary ==================================
✔ 6 checks passed
⚠ 1 warning
✘ 1 error
```

**Reply by polm** [reply]
OK, a couple of things could be going wrong.

One thing is that entities can't cross sentence boundaries. It looks like the number is getting picked up correctly, so the period there isn't a problem, but you might want to check that the colon isn't being treated as a sentence boundary, which is unlikely but possible.

Another thing is that it looks like because you have long documents, you might have enough labels, but the number of documents being low could be an issue. Since your documents are long maybe you can try breaking them up? Even if you have enough examples, batching related to the small number of documents will make your updates weird. 

About the entity spans including punctuation being a warning, I think you don't have to worry about that too much here. Entities for the default spaCy models don't include punctuation, and the warning is there for when you include periods at the end of a name or something. In your case it looks like you have colons in numbers, which shouldn't be an obstacle to learning.

In summary, I think this has to do with your small number of documents. I would try getting more training data or splitting your long documents into smaller ones, say a paragraph a piece or so.

**Reply by farrandi** [reply]
I tried your suggestion and I still got the same values. Here is the debug data:
```
============================ Data file validation ============================
✔ Corpus is loadable
✔ Pipeline can be initialized with data

=============================== Training stats ===============================
Language: en
Training pipeline: tok2vec, ner
400 training docs
400 evaluation docs
⚠ 324 training examples also in evaluation data
⚠ Low number of examples to train a new pipeline (400)

============================== Vocab & Vectors ==============================
ℹ 24848 total word(s) in the data (2267 unique)
ℹ No word vectors present in the package

========================== Named Entity Recognition ==========================
ℹ 12 label(s)
0 missing value(s) (tokens with '-' label)
⚠ 48 entity span(s) with punctuation
⚠ Low number of examples for label 'Location' (16)
⚠ Low number of examples for label 'Age' (17)
⚠ Low number of examples for label 'PHN' (14)
⚠ Low number of examples for label 'Patient' (49)
✔ Examples without occurrences available for all labels
✔ No entities consisting of or starting/ending with whitespace
Entity spans consisting of or starting/ending with punctuation can not be
trained with a noise level > 0.

================================== Summary ==================================
✔ 4 checks passed
⚠ 7 warnings
```
If this helps this is also my config.cfg file:
```
[paths]
train = "./training data/set 3/training.spacy"
dev = "./training data/set 3/validation.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_per_type = null
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
# before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

[initialize.before_init]
@callbacks = "customize_tokenizer"
```

and this is my tokenizer functions.py:
```python:
import re
from spacy.lang.char_classes import LIST_PUNCT, LIST_ELLIPSES, LIST_QUOTES, LIST_CURRENCY
from spacy.lang.char_classes import LIST_ICONS, HYPHENS, CURRENCY, UNITS
from spacy.lang.char_classes import CONCAT_QUOTES, ALPHA_LOWER, ALPHA_UPPER, ALPHA, PUNCT
from spacy.util import registry, compile_infix_regex, compile_prefix_regex, compile_suffix_regex
from spacy.tokenizer import Tokenizer
from spacy.symbols import ORTH

@registry.callbacks("customize_tokenizer")
def make_customize_tokenizer():
    def customize_tokenizer(nlp):
        
        split_chars = lambda char: list(char.strip().split(" "))

        prefixes = (
                ["§", "%", "=", "—", "–", r"\+(?![0-9])"]
                + split_chars( r"… …… , : ; \! \? ¿ ؟ ¡ \( \) \[ \] \{ \} < > _ # \ & 。 ？ ！ ， 、 ； ： ～ · । ، ۔ ؛ ٪") #removed [*]
                + LIST_ELLIPSES
                + LIST_QUOTES
                + LIST_CURRENCY
                + LIST_ICONS
            )
        prefix_re = compile_prefix_regex(prefixes)
        nlp.tokenizer.prefix_search = prefix_re.search


        # Modify tokenizer infix patterns
        infixes = (
            LIST_ELLIPSES
            + LIST_ICONS
            + [
                r"(?<=[0-9])[+\-\*^](?=[0-9-])",
                r"(?<=[{al}{q}])\.(?=[{au}{q}])".format(
                    al=ALPHA_LOWER, au=ALPHA_UPPER, q=CONCAT_QUOTES
                ),
                r"(?<=[{a}]),(?=[{a}])".format(a=ALPHA),
                r"(?<=[{a}])(?:{h})(?=[{a}])".format(a=ALPHA, h=HYPHENS),
                r"(?<=[{a}0-9])[:<>=/](?=[{a}])".format(a=ALPHA),
                r"(?<=[{a}0-9])[:/=](?=[{a}0-9])".format(a=ALPHA),  # added this to separate 'obs:30'
                r"(?<=[{a}])(?=[0-9])".format(a=ALPHA),             # to separate something123
                r"(?<=[0-9])(?=[{a}])".format(a=ALPHA),             # to separate 3mmHg
                r"(?<=[)])[?= {:}]",
            ]
        )
        infix_re = compile_infix_regex(infixes)
        nlp.tokenizer.infix_finditer = infix_re.finditer

        # Modify suffix pattern
        suffixes = (
            LIST_PUNCT
            + LIST_ELLIPSES
            + split_chars( r'\ " ” “ ` ‘ ´ ’ ‚ , „ » « 「 」 『 』 （ ） 〔 〕 【 】 《 》 〈 〉') # got rid of ['] in here
            + LIST_ICONS
            + ["'s", "'S", "’s", "’S", "—", "–"]
            + [
                r"(?<=[0-9])\+",
                r"(?<=°[FfCcKk])\.",
                r"(?<=[0-9])(?:{c})".format(c=CURRENCY),
                r"(?<=[0-9])(?:{u})".format(u=UNITS),
                r"(?<=[0-9{al}{e}{p}(?:{q})])\.".format(
                    al=ALPHA_LOWER, e=r"%²\-\+", q=CONCAT_QUOTES, p=PUNCT
                ),
                r"(?<=[{au}][{au}])\.".format(au=ALPHA_UPPER),
                r"(?<={a})"
            ]
        )
        suffix_re = compile_suffix_regex(suffixes)
        nlp.tokenizer.suffix_search = suffix_re.search

    return customize_tokenizer
```
I think the trained model does not generalize things well.

examples:
it would recognize `LV EF:3.12%` (even without whitespaces next to the colons)
but it wont recognize `LV FS:3.12%` (it would just recognize `LV` as Observation and `%` as Units and won't recognize the number `3,12` as Value.

Does this have something to do with data and overfitting?

**Reply by polm** [reply]
Thanks for the update, this makes several things clear. It looks like the main problem is you actually don't have enough data.

```
⚠ Low number of examples for label 'Location' (16)
⚠ Low number of examples for label 'Age' (17)
⚠ Low number of examples for label 'PHN' (14)
⚠ Low number of examples for label 'Patient' (49)
```

This is a serious problem - this is nowhere near enough examples and it's unsurprising your model can't generalize if this is all it has. (It is weird you didn't get this warning before, I'm not sure if that's a bug.)

```
⚠ 324 training examples also in evaluation data
```

This is also strange - is it intentional that you have this much overlap out of 400 docs? It's OK to use the same data for testing that things work, but for a real model your dev and train set should be separate.

Based on this it seems like the main thing you should do is focus on getting more data.

**Reply by polm** [reply]
Another thing, backing up a bit, I have some questions/suggestions about your entities. For Location and Patient, have you tried using the pretrained models? For PHN, which I guess is a patient number, is using regex an option? 

On reflection, since you don't have the low annotation warning about your Observation entity, you should have enough of those... Depending on how many examples you have overfitting could still be a concern. 

> but it wont recognize LV FS:3.12% (it would just recognize LV as Observation and % as Units and won't recognize the number 3,12 as Value.

Looking at what you're trying to label as an entity here, I'm not sure training an NER model is the best way to approach this. If you want to treat `3.12%` as an entity can't you just use a Matcher pattern with an EntityRuler to flag those? For example...

1. I assume that there's a fixed number of valid values for `UNITS` - maybe you have "lbs", "kgs", "kg", "%", or something, but if you can make a list of all those, using a rule-based matcher should be easier than NER. 
2. Once you match units, it should be easy to match any number in front of `UNITS` as a `VALUE` using rule based matching.

For the `LV FS` / `OBSERVATION` label, based on your one example I'm not sure, but is it possible to make a list of all possible valid values? For example, if I was going to pull US states out of text, I could make a list of all of those and use rule-based matching to find them, and that's easier than training a statistical model. But I can't list every valid person's name, which is why training an NER model is the right thing to do. 

If your observations are more like personal names than US states, then it sounds like you just need more examples and indeed overfitting is your problem, and you can work on fixing it with more data. But if they're more like states I would recommend you try writing a rule to match them.

**Reply by farrandi** [reply]
Oh That is my bad, I must have took the wrong dev set it should only be a 100, although currently, it has 29 overlappings (still gives the same value). Also for the separated ones I used less data so I do not think it is a bug. 

Thank you for the suggestion, will try it out and let you know!

**Reply by farrandi** [reply]
Is there a way to add a custom tokenizer to a model without training it? I am experimenting with using the list and making rule-based matcher for `UNITS` and `VALUE` but I am not sure how to add my custom tokenizer to the blank model.

**Reply by polm** [reply]
You can just call your `customize_tokenizer` function directly on the blank nlp object before you use it.

```
nlp = spacy.blank("en")
customize_tokenizer = make_customize_tokenizer()
customize_tokenizer(nlp)
```

**Reply by farrandi** [reply]
Hello Polm, 
I was experimenting with the rule-based matcher using units and values and it works when loaded from a blank model. 
I was wondering what would happen if we used the rule-based matcher from an existing model? would it override the entities?
right now it is giving me this error:
```
ValueError                                Traceback (most recent call last)
<ipython-input-22-44abbfe98511> in <module>
      4 # print(text)
      5 print("loaded file {x} ".format(x=number))
----> 6 doc = nlp(text)
      7 spacy.displacy.render(doc, style = "ent", jupyter = True, options = options)

~/.local/lib/python3.8/site-packages/spacy/language.py in __call__(self, text, disable, component_cfg)
    998                 raise ValueError(Errors.E109.format(name=name)) from e
    999             except Exception as e:
-> 1000                 error_handler(name, proc, [doc], e)
   1001             if doc is None:
   1002                 raise ValueError(Errors.E005.format(name=name))

~/.local/lib/python3.8/site-packages/spacy/util.py in raise_error(proc_name, proc, docs, e)
   1501 
   1502 def raise_error(proc_name, proc, docs, e):
-> 1503     raise e
   1504 
   1505 

~/.local/lib/python3.8/site-packages/spacy/language.py in __call__(self, text, disable, component_cfg)
    993                 error_handler = proc.get_error_handler()
    994             try:
--> 995                 doc = proc(doc, **component_cfg.get(name, {}))
    996             except KeyError as e:
    997                 # This typically happens if a component is not initialized

<ipython-input-21-798a595f7c41> in value_entity(doc)
     16         else:
     17             new_ents.append(ent)
---> 18     doc.ents = new_ents
     19     return doc
     20 nlp.add_pipe("value_entity")

~/.local/lib/python3.8/site-packages/spacy/tokens/doc.pyx in spacy.tokens.doc.Doc.ents.__set__()

~/.local/lib/python3.8/site-packages/spacy/tokens/doc.pyx in spacy.tokens.doc.Doc.set_ents()

ValueError: [E1010] Unable to set entity information for token 825 which is included in more than one span in entities, blocked, missing or outside.
```

This is the code I added:
```python:
# make a list of matchers for UNITS
LIST_OF_UNITS = ["cm","cm 2","cm2","m","m2", "m 2", "cm/m 2", "cm / m 2", "l/min", "l / min", "l/m*m 2", "m/s", "m/s"
                 "grams/m2", "grams/m 2", "g/m 2","g/m2","ml/m2", "ml/m 2", "cm/sec", "cm2/m2", "cm 2/m 2", "kg/m 2",
                 "kg/m2",
                 "%", "mmHg", "mmhg","mm Hg", "msec", "ml", "pounds", "lbs", "inches"]

ruler = nlp.add_pipe("entity_ruler")
patterns = [{"label":"UNITS", "pattern": units} for units in LIST_OF_UNITS]
ruler.add_patterns(patterns)

# using rule based matching for VALUE
@Language.component("value_entity")
def value_entity(doc):
    new_ents = []
    for ent in doc.ents:
        if ent.label_ == "UNITS" and ent.start != 0:
            prev_token = doc[ent.start - 1]
            
            expression = r"[0-9]*(.)[0-9]*"
            if re.match(expression, prev_token.text):
                new_ent = Span(doc, ent.start - 1, ent.start, label="VALUE")
                new_ents.append(new_ent)
                new_ent = Span(doc, ent.start, ent.end, label=ent.label_)
                new_ents.append(new_ent)
        else:
            new_ents.append(ent)
    doc.ents = new_ents
    return doc
nlp.add_pipe("value_entity")
```

Thank you!

**Reply by polm** [reply]
Whether it overwrites or not is an option, see [the docs](https://spacy.io/usage/rule-based-matching#:~:text=to%20overwrite%20overlapping%20entities). 

The trained NER components do not overwrite existing entities, so the order you put them in the pipeline is important.
